bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/RowResultAdapter.java

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.hbase.wrappers.veneer;
+
+import com.google.api.core.InternalApi;
+import com.google.cloud.bigtable.data.v2.models.RowAdapter;
+import com.google.cloud.bigtable.hbase.adapters.read.RowCell;
+import com.google.cloud.bigtable.hbase.util.ByteStringer;
+import com.google.cloud.bigtable.hbase.util.TimestampConverter;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.client.Result;
+
+/**
+ * Adapter for {@link RowAdapter} that uses {@link Result} to represent logical rows.
+ *
+ * <p>For internal use only - public for technical reasons.
+ */
+@InternalApi("For internal usage only")
+public class RowResultAdapter implements RowAdapter<Result> {
+
+  private static final byte[] EMPTY_VALUE = new byte[0];
+
+  @Override
+  public RowBuilder<Result> createRowBuilder() {
+    return new RowResultBuilder();
+  }
+
+  @Override
+  public boolean isScanMarkerRow(Result result) {
+    if (result instanceof RowResult) {
+      return ((RowResult) result).isMarkerRow();
+    }
+    // TODO: This may never be executed still is it ok to fallback on this?
+    return result.isEmpty();
+  }
+
+  @Override
+  public ByteString getKey(Result result) {
+    if (result instanceof RowResult) {
+      return ((RowResult) result).getKey();
+    }
+    return ByteStringer.wrap(result.getRow());
+  }
+
+  static class RowResult extends Result {
+    private final ByteString rowKey;
+    private final boolean isMarkerRow;
+
+    static RowResult create(ByteString rowKey, List<Cell> cells) {
+      return new RowResult(rowKey, cells);
+    }
+
+    static RowResult createMarker(ByteString rowKey) {
+      return new RowResult(rowKey, ImmutableList.<Cell>of());
+    }
+
+    private RowResult(ByteString rowKey, List<Cell> cells) {
+      this.rowKey = rowKey;
+      this.isMarkerRow = cells == null || cells.isEmpty();
```

Comment:
Can this be explicitly passed from the factory methods?

Reply:
Yes, we could another param to RowResult ctor, 
```java
private RowResult(ByteString rowKey, boolean isMarkerRow, List<Cell> cells) {
      this.rowKey = rowKey;
      this.isMarkerRow = isMarkerRow;
      // ...
    }
```
May I suggest an alternative, Looking at finishRow(), `List<Cell>` would never be `null`, then we can directly use that to check for markerRow, eg:
```java
private RowResult(ByteString rowKey, List<Cell> cells) {
      this.rowKey = rowKey;
      this.isMarkerRow = cells.isEmpty();
      // ...
    }
```
Please let me know your thoughts on this:

Reply:
Have updated with `isMarkerRow` as a `ctor` param, this would save us a check after each Row creation(looked up StateMachine.java to understand the usages).